cpp/src/jni/orc/concurrent_map.h

```diff
@@ -0,0 +1,87 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ */
+
+#ifndef JNI_ID_TO_MODULE_MAP_H
+#define JNI_ID_TO_MODULE_MAP_H
+
+#include <memory>
+#include <mutex>
+#include <unordered_map>
+#include <utility>
+
+#include "arrow/util/macros.h"
+
+namespace arrow {
+namespace jni {
+
+/**
+ * An utility class that map module id to module pointers.
+ * @tparam Holder class of the object to hold.
+ */
+template <typename Holder>
+class concurrentMap {
+ public:
+  concurrentMap() : module_id_(kInitModuleId) {}
+
+  jlong Insert(Holder holder) {
+    std::lock_guard<std::mutex> lock(mtx_);
+    jlong result = module_id_++;
+    map_.insert(std::pair<jlong, Holder>(result, holder));
+    return result;
+  }
+
+  void Erase(jlong module_id) {
+    std::lock_guard<std::mutex> lock(mtx_);
+    map_.erase(module_id);
+  }
+
+  Holder Lookup(jlong module_id) {
+    Holder result = NULLPTR;
+    try {
+      result = map_.at(module_id);
+    } catch (const std::out_of_range& e) {
+    }
+    if (result != NULLPTR) {
+      return result;
+    }
+    std::lock_guard<std::mutex> lock(mtx_);
+    try {
+      result = map_.at(module_id);
+    } catch (const std::out_of_range& e) {
+    }
+    return result;
+  }
+
+  void Clear() {
+    std::lock_guard<std::mutex> lock(mtx_);
+    map_.clear();
+  }
+
+ private:
+  // starting value of the module_id.
```

Comment:
I'm still not sure I understand this, why does module ID start at 4?

Reply:
The concurrent_map implementation is based on the design of its Gandiva [counterpart](https://github.com/dremio/gandiva/blob/master/cpp/src/jni/id_to_module_map.h)

For my understand, this is a random magic number to avoid the module ID start from zero for certain pitfalls.

Reply:
@pravindra @praveenbingo?

Reply:
yeah, i think we wanted it to be non-zero to help debug the cases where the caller is using an  uninitialized variable from java.

Reply:
that sounds reasonable.  @yuruiz please add this to the documentation.